### PR TITLE
feat: decode sequencer batch

### PR DESCRIPTION
**Pull request by tynes**
There was an off by one error for the batch submitter where it would submit an extra element in each batch. The fix was to remove a `+ 1` that came with a comment regarding add one because it was exclusive. Maybe previously it was exclusive? I observe the correct behavior with this path, the `MAX_BATCH_SIZE` will result in the correct value instead of it being greater by one. This also adds a `decodeSequencerBatchAppend` function that is useful for decoding calldata.

The tests had to be updated due to the off by one error

### CODE REVIEW

**Comment by karlfloersch** [comment]
Amazing thank you sir 😅

File: src/transaciton-chain-contract.ts [hunk 1]
```diff
@@ -86,3 +86,51 @@ const encodeBatchContext = (context: BatchContext): string => {
...
   )
 }
+
+export const decodeAppendSequencerBatch = (b: string): AppendSequencerBatchParams => {
+  b = remove0x(b)
+
+  const shouldStartAtBatch = b.slice(0, 10)
+  const totalElementsToAppend = b.slice(10, 16)
+
+  //const contextHeader = b.slice(16, 22)
```
Comment (tynes) [inline src/transaciton-chain-contract.ts:96]: this is dead code

File: src/transaciton-chain-contract.ts [hunk 2]
```diff
@@ -86,3 +86,49 @@ const encodeBatchContext = (context: BatchContext): string => {
...
+    })
+  }
+
+  const transactions = []
+  for (const context of contexts) {
+    for (let i = 0; i < context.numSequencedTransactions; i++) {
+      const size = b.slice(offset, offset + 6)
+      offset += 6
+      const raw = b.slice(offset, offset + (parseInt(size, 16) * 2))
+      transactions.push(raw)
```
Comment (tynes) [inline src/transaciton-chain-contract.ts:123]: Maybe these should have a `0x` prefix

File: test/batch-submitter/batch-encoder.spec.ts [hunk 3]
```diff
@@ -40,4 +40,35 @@ describe('BatchEncoder', () => {
...
+      const batch = {
+        shouldStartAtBatch: 10,
+        totalElementsToAppend: 1,
+        contexts: [{
+          numSequencedTransactions: 2,
+          numSubsequentQueueTransactions: 1,
+          timestamp: 100,
+          blockNumber: 200
+        }],
+        transactions: ['45423400000011', '45423400000012'],
```
Comment (tynes) [inline test/batch-submitter/batch-encoder.spec.ts:67]: these need a 0x

**Comment by tynes** [comment]
Ok I think this is ready for a final review/merge

File: src/batch-submitter/tx-batch-submitter.ts [hunk 4]
```diff
@@ -157,14 +157,13 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
   }
 
   public async _getBatchStartAndEnd(): Promise<Range> {
-    // TODO: Remove BLOCK_OFFSET by adding a tx to Geth's genesis
```
Comment (tynes) [inline src/batch-submitter/tx-batch-submitter.ts:160]: We are past this being a thing so I think its ok removing the comment

File: src/batch-submitter/tx-batch-submitter.ts [hunk 5]
```diff
@@ -157,14 +157,13 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
...
     const startBlock =
       (await this.chainContract.getTotalElements()).toNumber() + BLOCK_OFFSET
     const endBlock =
       Math.min(
         startBlock + this.maxBatchSize,
         await this.l2Provider.getBlockNumber()
-      ) + 1 // +1 because the `endBlock` is *exclusive*
+      ) + 1 // +1 because the endBlock is *exclusive*
```
Comment (tynes) [inline src/batch-submitter/tx-batch-submitter.ts:166]: This is an unnecessary diff
